### PR TITLE
[DOCS] Move string manipulation FAQ items to type docs

### PR DESCRIPTION
**Pull request by ChrisChinchilla**
It seems questions on string manipulation come up a lot, so I moved the FAQ item to the main docs and consolidated it. Part of https://github.com/ethereum/solidity/issues/1219

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] README / documentation was extended, if necessary
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### CODE REVIEW

File: docs/types/value-types.rst [hunk 1]
```diff
@@ -453,6 +453,13 @@ String literals are written with either double or single-quotes (``"foo"`` or ``
 
 For example, with ``bytes32 samevar = "stringliteral"`` the string literal is interpreted in its raw byte form when assigned to a ``bytes32`` type.
 
+.. note::
+
+  While Solidity does not have string manipulation functions, you can use
+  this implicit conversion for equivalent functionality. For example to compare
+  two strings ``keccak256(abi.encode(s1)) == keccak256(abi.encode(s2))``, or to
+  concatenate two strings already encoded with ``abi.encodePacked(a, b);``.
```
Comment (erak) [inline docs/types/value-types.rst]:
```suggestion
  concatenate two strings already encoded with ``abi.encodePacked(s1, s2);``.
```

File: docs/types/value-types.rst [hunk 2]
```diff
@@ -453,6 +453,13 @@ String literals are written with either double or single-quotes (``"foo"`` or ``
...
+
```
Comment (erak) [inline docs/types/value-types.rst]:
```suggestion
In general, you should always convert strings to a ``bytes`` before manipulating it.
```
This way people could possibly derive something like the append functionality on their own ;)

**Comment by ChrisChinchilla** [comment]
Thank @bit-shift fixed a little formatting, could you review again? :)

**Review by erak — APPROVED** [review]

**Review by erak — CHANGES_REQUESTED** [review]
Oh wait, the tests are failing.

**Comment by ekpyron** [comment]
@bit-shift Yeah - the tests are a bit unstable at the moment... someone should probably look into it :-D (I just created #5803 for tracking this) - but it's not related to this PR.

**Comment by erak** [comment]
@ekpyron Thanks for letting me know ;) The failure was indeed not related to the docs changes.

**Review by erak — DISMISSED** [review]

**Comment by ChrisChinchilla** [comment]
And there was a merge conflict, one more review @bit-shift ;)

File: docs/types/value-types.rst [hunk 3]
```diff
@@ -453,6 +453,15 @@ String literals are written with either double or single-quotes (``"foo"`` or ``
 
 For example, with ``bytes32 samevar = "stringliteral"`` the string literal is interpreted in its raw byte form when assigned to a ``bytes32`` type.
 
+.. note::
+
+  While Solidity does not have string manipulation functions, you can use
+  this implicit conversion for equivalent functionality. For example to compare
+  two strings ``keccak256(abi.encode(s1)) == keccak256(abi.encode(s2))``, or to
```
Comment (axic) [inline docs/types/value-types.rst]: The first sentence talks about conversion, but the next is about comparison.

Comment (ChrisChinchilla) [reply docs/types/value-types.rst]: @axic Sure, but as far as I have been able to tell, you have to convert to be able to do the comparison?

File: docs/control-structures.rst [hunk 4]
```diff
@@ -264,6 +264,32 @@ Complications for Arrays and Structs
 The semantics of assignments are a bit more complicated for non-value types like arrays and structs.
 Assigning *to* a state variable always creates an independent copy. On the other hand, assigning to a local variable creates an independent copy only for elementary types, i.e. static types that fit into 32 bytes. If structs or arrays (including ``bytes`` and ``string``) are assigned from a state variable to a local variable, the local variable holds a reference to the original state variable. A second assignment to the local variable does not modify the state but only changes the reference. Assignments to members (or elements) of the local variable *do* change the state.
 
+
+In the example below the call to ``g(x)`` has no effect on ``x`` because it needs
```
Comment (chriseth) [inline docs/control-structures.rst]: `needs to create` -> `creates`.

File: docs/types/value-types.rst [hunk 5]
```diff
@@ -453,6 +453,15 @@ String literals are written with either double or single-quotes (``"foo"`` or ``
...
+  concatenate two strings already encoded with ``abi.encodePacked(s1, s2);``.
+
+  In general, you should always convert strings to a ``bytes`` before manipulating it.
```
Comment (chriseth) [inline docs/types/value-types.rst]: No way to `should` here, you have to.

File: docs/control-structures.rst (hunk 4 above)
Comment (ChrisChinchilla) [reply docs/control-structures.rst]: @chriseth That sounds like an edit I would do ;) Must have missed it in the back-and-forth ;)

File: docs/control-structures.rst [hunk 6]
```diff
@@ -264,13 +264,13 @@ Complications for Arrays and Structs
 The semantics of assignments are a bit more complicated for non-value types like arrays and structs.
 Assigning *to* a state variable always creates an independent copy. On the other hand, assigning to a local variable creates an independent copy only for elementary types, i.e. static types that fit into 32 bytes. If structs or arrays (including ``bytes`` and ``string``) are assigned from a state variable to a local variable, the local variable holds a reference to the original state variable. A second assignment to the local variable does not modify the state but only changes the reference. Assignments to members (or elements) of the local variable *do* change the state.
 
-In the example below the call to ``g(x)`` has no effect on ``x`` because it needs
-to create an independent copy of the storage value in memory. However ``h(x)`` modifies ``x`` because a reference and
-not a copy is passed.
+In the example below the call to ``g(x)`` has no effect on ``x`` because it creates
+an independent copy of the storage value in memory. However, ``h(x)`` successfully modifies ``x``
```
Comment (chriseth) [inline docs/control-structures.rst:268]: Ah, that reads really nicely now!

File: docs/types/value-types.rst [hunk 7]
```diff
@@ -456,6 +456,15 @@ String literals are written with either double or single-quotes (``"foo"`` or ``
 
 For example, with ``bytes32 samevar = "stringliteral"`` the string literal is interpreted in its raw byte form when assigned to a ``bytes32`` type.
 
+.. note::
+
+  While Solidity does not have string manipulation functions, you can use
```
Comment (chriseth) [inline docs/types/value-types.rst]: Actually I think this note should rather go to string types and not to string literal types.

Comment (ChrisChinchilla) [reply docs/types/value-types.rst]: @chriseth But there is no such section (anymore?), or do you just mean further down this section?

Comment (chriseth) [reply docs/types/value-types.rst]: https://solidity.readthedocs.io/en/develop/types.html#arrays - we could introduce a small header there.

Comment (ChrisChinchilla) [reply docs/types/value-types.rst]: OK done, I'm going to revisit this for some polish later, as I feel strings are a bit confusing for some…

**Review by erak — APPROVED** [review]

File: docs/types/reference-types.rst [hunk 8]
```diff
@@ -106,13 +106,24 @@ Array elements can be of any type, including mapping or struct. The general
 restrictions for types apply, in that mappings can only be stored in the
 ``storage`` data location and publicly-visible functions need parameters that are :ref:`ABI types <ABI>`.
 
+It is possible to mark arrays ``public`` and have Solidity create a :ref:`getter <visibility-and-getters>`.
```
Comment (chriseth) [inline docs/types/reference-types.rst:109]: I don't think this is a good place to put it. This section is about arrays as concepts of types, not about state variables. Put here, it reads like you could add `public` to any array type.

File: docs/types/reference-types.rst [hunk 9]
```diff
@@ -106,13 +106,24 @@ Array elements can be of any type, including mapping or struct. The general
...
 
+``bytes`` and ``strings`` as Arrays
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
 Variables of type ``bytes`` and ``string`` are special arrays. A ``bytes`` is similar to ``byte[]``,
 but it is packed tightly in calldata and memory. ``string`` is equal to ``bytes`` but does not allow
 length or index access.
 
+While Solidity does not have string manipulation functions, you can use
+this implicit conversion for equivalent functionality. For example to compare
```
Comment (chriseth) [inline docs/types/reference-types.rst:123]: Which implicit conversion?

File: docs/types/reference-types.rst (hunk 8 above)
Comment (chriseth) [reply docs/types/reference-types.rst:109]: Ah, I see this was only moved. So in this case, we should clarify that it only works on state variables.

Comment (ChrisChinchilla) [reply docs/types/reference-types.rst]: @chriseth Previously this was at the bottom of the section, also in a place that didn't make much sense. I have amended, but there are other examples of text very similar to this in other places in the docs, where it's equally unclear, so I might do a clarification around all of those.

**Pull request merged**